cpp/src/arrow/adapters/orc/adapter_util.cc

```diff
@@ -315,13 +342,462 @@ Status AppendBatch(const liborc::Type* type, liborc::ColumnVectorBatch* batch,
+      return WriteFixedSizeBinaryBatch(column_vector_batch, arrow_offset, orc_offset,
+                                       length, array, incoming_mask);
+    case arrow::Type::type::DATE32:
+      return WriteNumericBatch<arrow::NumericArray<arrow::Date32Type>,
+                               liborc::LongVectorBatch, int64_t>(
+          column_vector_batch, arrow_offset, orc_offset, length, array, incoming_mask);
+    case arrow::Type::type::DATE64:
+      return WriteTimestampBatch<arrow::Date64Array>(
+          column_vector_batch, arrow_offset, orc_offset, length, array, incoming_mask,
+          kOneSecondMillis, kOneMilliNanos);
+    case arrow::Type::type::TIMESTAMP: {
+      switch (arrow::internal::checked_pointer_cast<arrow::TimestampType>(array.type())
+                  ->unit()) {
+        case arrow::TimeUnit::type::SECOND:
+          return WriteTimestampBatch<arrow::TimestampArray>(
+              column_vector_batch, arrow_offset, orc_offset, length, array, incoming_mask,
+              1, kOneSecondNanos);
+        case arrow::TimeUnit::type::MILLI:
+          return WriteTimestampBatch<arrow::TimestampArray>(
+              column_vector_batch, arrow_offset, orc_offset, length, array, incoming_mask,
+              kOneSecondMillis, kOneMilliNanos);
+        case arrow::TimeUnit::type::MICRO:
+          return WriteTimestampBatch<arrow::TimestampArray>(
+              column_vector_batch, arrow_offset, orc_offset, length, array, incoming_mask,
+              kOneSecondMicros, kOneMicroNanos);
+        default:
+          return WriteTimestampBatch<arrow::TimestampArray>(
+              column_vector_batch, arrow_offset, orc_offset, length, array, incoming_mask,
+              kOneSecondNanos, 1);
+      }
+    }
+    case arrow::Type::type::DECIMAL:
+      return WriteDecimalBatch(column_vector_batch, arrow_offset, orc_offset, length,
+                               array, incoming_mask);
+    case arrow::Type::type::STRUCT:
+      return WriteStructBatch(column_vector_batch, arrow_offset, orc_offset, length,
+                              array, incoming_mask);
+    case arrow::Type::type::LIST:
+      return WriteListBatch<arrow::ListArray>(column_vector_batch, arrow_offset,
+                                              orc_offset, length, array, incoming_mask);
+    case arrow::Type::type::LARGE_LIST:
+      return WriteListBatch<arrow::LargeListArray>(
+          column_vector_batch, arrow_offset, orc_offset, length, array, incoming_mask);
+    case arrow::Type::type::FIXED_SIZE_LIST:
+      return WriteListBatch<arrow::FixedSizeListArray>(
+          column_vector_batch, arrow_offset, orc_offset, length, array, incoming_mask);
+    case arrow::Type::type::MAP:
+      return WriteMapBatch(column_vector_batch, arrow_offset, orc_offset, length, array,
+                           incoming_mask);
+    default: {
+      return arrow::Status::Invalid("Unknown or unsupported Arrow type kind: ", kind);
```

Comment:
Using the array's type->ToString() would give a better error.

Reply:
Thanks! Fixed!